src/components/UI/cluster_dashboard_item.js

```diff
@@ -0,0 +1,260 @@
+import {
+  clusterNodePools,
+  getCpusTotal,
+  getCpusTotalNodePools,
+  getMemoryTotal,
+  getMemoryTotalNodePools,
+  getNumberOfNodePoolsNodes,
+  getNumberOfNodes,
+  getStorageTotal,
+} from 'utils/cluster_utils';
+import { Dot } from 'styles';
+import { Link } from 'react-router-dom';
+import { relativeDate } from 'lib/helpers.js';
+import Button from 'UI/button';
+import ButtonGroup from 'react-bootstrap/lib/ButtonGroup';
+import ClusterIDLabel from 'UI/cluster_id_label';
+import EmptyStateDisplay from 'UI/empty_state_display';
+import moment from 'moment';
+import PropTypes from 'prop-types';
+import React from 'react';
+import RefreshableLabel from 'UI/refreshable_label';
+import styled from '@emotion/styled';
+
+const Wrapper = styled.div`
+  display: flex;
+
+  b {
+    font-weight: 400;
+  }
+`;
+
+const Label = styled.div`
+  flex: 0 0 90px;
+  font-size: 1.2em;
+`;
+
+const Content = styled.div`
+  flex: 1;
+  padding-right: 15px;
+`;
+
+const Title = styled.div`
+  font-size: 1.2em;
+  z-index: 120;
+  position: relative;
+
+  span {
+    font-weight: bold;
+  }
+`;
+
+const Buttons = styled.div`
+  text-align: right;
+  flex: 0 0 210px;
+
+  .progress_button--container:last-child {
+    margin-right: 0px;
+  }
+`;
+
+class ClusterDashboardItem extends React.Component {
+  state = {
+    enforceReRender: null,
+    nodePools: [],
+  };
+
+  componentDidMount() {
+    this.registerReRenderInterval();
+
+    const { cluster, nodePools } = this.props;
+
+    if (this.props.isNodePool) {
+      const nodePoolsData = clusterNodePools(nodePools, cluster);
+      this.setState({ nodePools: nodePoolsData });
+    }
+  }
+
+  componentWillUnmount() {
+    window.clearInterval(this.reRenderInterval);
+  }
+
+  /**
+   * Activates periodic re-rendering to keep displayed info, like relative
+   * dates, fresh.
+   */
+  registerReRenderInterval = () => {
+    var refreshInterval = 10 * 1000; // 10 seconds
+    this.reRenderInterval = window.setInterval(() => {
+      // enforce re-rendering by state change
+      this.setState({ enforceReRender: Date.now() });
+    }, refreshInterval);
+  };
+
+  /**
+   * Returns true if the cluster is younger than 30 days
+   */
+  clusterYoungerThan30Days() {
+    var age = Math.abs(
+      moment(this.props.cluster.create_date)
+        .utc()
+        .diff(moment().utc()) / 1000
+    );
+
+    return age < 30 * 24 * 60 * 60;
+  }
+
+  /**
+   * Returns the amount of memory in this cluster.
+   */
+  memory() {
```

Comment:
this and other function like `cpus` `storage` and `numNodes` used to be procedural code at the start of the `render()` method.

I think we should try to keep the render method as clean as possible, so i moved it out into individual functions.

It does mean we call these functions about 2 more times (since they are used in two places in the render method) but if that really becomes a performance problem, it would be easy to memoize this function.

But for now let's not premature optimise?

Reply:
Instead memoization or as a way of doing it, let's use state. That's what it is for. We can execute this only when `cluster` prop change, no need to check for deep nested properties thanks to Immer.